### DISCUSSION TITLE
How can I use something like codesnippets in my markdown files?

### DISCUSSION THREAD

**Discussion by marcofranssen** [vercel/next.js#20491] [Help]
I'm trying to migrate my blog from https://hexo.io/ to nextjs.

In hexo I'm using following type of snippets to include them as codeblocks in my blog from external files.

```
{% include_code App.js lang:jsx react-router-nginx-http2/src/App.js %}
```

They render as other backtick type of codeblocks at build time during static page generation. See here a result of such code-block.

https://marcofranssen.nl/building-a-elasticsearch-cluster-using-docker-compose-and-traefik/#Summary

Similarly Hugo supports this feature as well via `blocks` https://gohugo.io/templates/base/

Now I'm really struggling to get this done in nextjs. I followed the blog example so far.

I have tried by replacing with the following and fetch the file contents at runtime (although this is less ideal as statically rendering them).

```
<CodeBlock language="jsx" title="App.jsx" url="react-router-nginx-http2/src/App.js" />
```

For that I build this component. 

```js
import { useState, useEffect } from "react";
import Highlight, { defaultProps } from "prism-react-renderer";

export default function CodeBlock({ title, url, language }) {
  const [code, setCode] = useState("");
  const [isLoading, setIsLoading] = useState("");

  useEffect(() => {
    const fetchCode = async () => {
      setIsLoading(true);
      const res = await fetch(`/downloads/code/${url}`);
      const body = await res.text();
      setCode(body);
      setIsLoading(false);
    };

    fetchCode();
  }, []);

  return isLoading ? (
    <div className="prose remark-highlight">
      <pre className={`language-${language}`}>
        <code className={`language-${language}`}>Loading...</code>
      </pre>
    </div>
  ) : (
    <Highlight
      {...defaultProps}
      theme={undefined}
      code={code}
      language={language}
    >
      {({ className, tokens, getLineProps, getTokenProps }) => (
        <div className="prose remark-highlight">
          <pre className={className}>
            <code className={className}>
              {tokens.map((line, i) => (
                <div {...getLineProps({ line, key: i })}>
                  {line.map((token, key) => (
                    <span {...getTokenProps({ token, key })} />
                  ))}
                </div>
              ))}
            </code>
          </pre>
        </div>
      )}
    </Highlight>
  );
}
```

Although this is less ideal as it would require the page to fetch this block at runtime, I also don't manage to get it working. when I use these blocks in my `_post/my-blog.md`.

When trying this approach https://github.com/vercel/next.js/tree/canary/examples/with-mdx-remote I get following error.

```bash
./node_modules/jsdom/lib/jsdom/living/xhr/XMLHttpRequest-impl.js:4:19
Module not found: Can't resolve 'child_process'
null
```

Would really appreciate some help here. This would be a killer feature for next-js that allows bloggers to write their blogs in markdown and leverage some snippets like this to load code examples from external files. Or to include other kind of widgets in the blog contents.

I also was searching if there is maybe a remark plugin that could scan the markdown for some reusable codeblocks or something like this but didn't find anything like this. That would for example allow to statically render all those references from the markdown.

**Comment by leerob** [comment]
I'd recommend checking out MDX for this. It will be your replacement for templating with vanilla Markdown.

**Reply by marcofranssen** [reply]
I did try that.

```
./node_modules/jsdom/lib/jsdom/living/xhr/XMLHttpRequest-impl.js:4:19
Module not found: Can't resolve 'child_process'
null
```

When trying to implement this example.

https://github.com/vercel/next.js/tree/canary/examples/with-mdx-remote

**Reply by marcofranssen** [reply]
It turns out I had to change a couple of more things in the markdown files themself to get this issue resolved. If I manage to get everything accordingly migrated I'm planning to capture my learnings in a blog. 🤞
